openmmmcmc/states.py

```diff
@@ -0,0 +1,1903 @@
+        LangevinIntegrator).
+
+        It can be set only to a system which is consistent with the
+        current thermodynamic state. Use set_system() if you want to
+        correct the thermodynamic state of the system automatically
+        before assignment.
+
+        See Also
+        --------
+        ThermodynamicState.get_system
+        ThermodynamicState.set_system
+        ThermodynamicState.create_context
+
+        Examples
+        --------
+
+        """
+        return copy.deepcopy(self._system)
+
+    @system.setter
+    def system(self, value):
+        self._check_system_consistency(value)
+        self._system = copy.deepcopy(value)
+        self._cached_standard_system_hash = None  # Invalidate cache.
+
+    def set_system(self, system, fix_state=False):
+        """Manipulate and set the system.
+
+        With default arguments, this is equivalent to using the system
+        property, which raises an exception if the thermostat and the
+        barostat are not configured according to the thermodynamic state.
+        With this method it is possible to adjust temperature and
+        pressure of the system to make the assignment possible, without
+        manually configuring thermostat and barostat.
+
+        Parameters
+        ----------
+        system : simtk.openmm.System
+            The system to set.
+        fix_state : bool, optional
+            If True, a thermostat is added to the system (if not already
+            present) and set to the correct temperature. If this state is
+            in NPT ensemble, a barostat is added or configured if it
+            exist already. If False, this simply check that thermostat
+            and barostat are correctly configured without modifying them.
+            Default is False.
+
+        Raises
+        ------
+        ThermodynamicsError
+            If the system after the requested manipulation is still in
+            an incompatible state.
+
+        Examples
+        --------
+        The constructor adds a thermostat and a barostat to configure
+        the system in an NPT ensemble.
+
+        >>> from openmmtools import testsystems
+        >>> alanine = testsystems.AlanineDipeptideExplicit()
+        >>> state = ThermodynamicState(alanine.system, temperature=300*unit.kelvin,
+        ...                            pressure=1.0*unit.atmosphere)
+
+        If we try to set a system not in NPT ensemble, an error occur.
+
+        >>> state.system = alanine.system
+        Traceback (most recent call last):
+        ...
+        ThermodynamicsError: System does not have a thermostat specifying the temperature.
+
+        We can fix both thermostat and barostat while setting the system.
+        >>> state.set_system(alanine.system, fix_state=True)
+
+        """
+        system = copy.deepcopy(system)
+        if fix_state:
+            self._set_system_thermostat(system, self.temperature)
```

Comment:
The consistency problem I run into here is that the property `system` guarantees the system to be in a specific thermodynamic state, which means that it needs a thermostat in the general case (it's still possible to strip it with `get_system(remove_thermostat=True)`. At the same time we don't want to have both a system thermostat and a thermostated integrator. I couldn't find a better solution than this one, but I'm very open to suggestions.